src/python/bot/minimizer/js_minimizer.py

```diff
@@ -30,47 +37,74 @@ def _execute(self, data):
 
     brace_stack = []
     paren_stack = []
-    last_start_of_line_token = 0
 
     for index, token in enumerate(testcase.tokens):
       if token == '{':
-        brace_stack.append((last_start_of_line_token, index))
+        brace_stack.append(index)
 
       elif token == '}' and brace_stack:
-        previous_line_start, previous_end = brace_stack.pop()
-
-        # Move |previous_line_start| to the first non-empty string.
-        while previous_line_start > 0:
-          if testcase.tokens[previous_line_start].strip():
-            break
-          previous_line_start -= 1
 
         # Two hypotheses for tokens grouped by curly braces:
-        # 1) Remove from start of line to curly brace and the closing brace.
+        # 1) Remove from start of line to open brace and the closing brace.
         #    e.g.: if (statement_that_evaluates_to_true) { crash() } -> crash()
-        hypothesis = list(range(previous_line_start,
-                                previous_end + 1)) + [index]
+        open_brace_index = brace_stack.pop()
+
+        # Find the first non-empty token prior to the starting brackets.
+        token_before_bracket = step_back_while(
+            open_brace_index - 1, (lambda x: not testcase.tokens[x].strip()))
+
+        # If that token is a close paren, we need to grab everything else too.
+        # Do this to grab the whole paren so we don't create a syntax error by
+        # removing only part of a paren.
+        if testcase.tokens[token_before_bracket] == ")":
+          # Find everything in the paren.
+          token_before_bracket = step_back_while(
+              token_before_bracket, (lambda x: testcase.tokens[x] != "("))
+
+          # Get the token before the paren.
+          token_before_bracket -= 1
+          token_before_bracket = step_back_while(
+              token_before_bracket, (lambda x: not testcase.tokens[x].strip()))
+
+        # Walk back to the start of that line as well to get if/else and funcs.
+        # Do this after paren to manage situations where there are newlines in
+        # the parens.
+        token_before_bracket = step_back_while(
+            token_before_bracket, (lambda x: testcase.tokens[x] != "\n"))
+
+        token_before_bracket += 1
+
+        hypothesis = list(range(token_before_bracket,
+                                open_brace_index + 1)) + [index]
+
         testcase.prepare_test(hypothesis)
 
         # 2) Remove previous tokens and from the closing brace to the next one.
-        #    e.g.: try { crash() } catch(e) {} -> crash()
+        #    e.g.: try { crash() } catch(e) {} -> crash().
         future_index = len(testcase.tokens)
+        open_count = 0
         for future_index in range(index + 1, len(testcase.tokens)):
+          if testcase.tokens[future_index] == "{":
```

Comment:
single quotes here and in all other places too. I thought `python butler.py format` used to catch and fix these, weird...